lib/lifeguard/threadpool.rb

```diff
@@ -63,64 +55,38 @@ def async(*args, &block)
       end
 
       @mutex.synchronize do
-        prune_busy_threads_without_mutex
-
         if busy_size < pool_size
           queued_the_work = true
 
-          @busy_threads << ::Thread.new(block, args, self) do |callable, call_args, parent|
-            begin
-              ::Thread.current[:__start_time_in_seconds__] = Time.now.to_i
-              ::Thread.current.abort_on_exception = false
-              callable.call(*call_args) # should we check the args? pass args?
-            ensure
-              parent.on_thread_exit(::Thread.current)
-            end
-          end
+          @busy_threads.add ::Thread.new(block, args, self) { |callable, call_args, parent|
+            ::Thread.current[:__start_time_in_seconds__] = Time.now.to_i
+            ::Thread.current.abort_on_exception = false
+            callable.call(*call_args) # should we check the args? pass args?
+          }
         end
 
-        prune_busy_threads_without_mutex
         queued_the_work
       end
     end
 
-    def prune_busy_threads
-      @mutex.synchronize do
-        prune_busy_threads_without_mutex
-      end
-    end
+    def shutdown(shutdown_timeout = 3)
+      kill_at = Time.now.to_f + shutdown_timeout
 
-    def shutdown(shutdown_timeout = 0)
       @mutex.synchronize do
-        prune_busy_threads_without_mutex
-
-        if @busy_threads.size > 0
-          # Cut the shutdown_timeout by 10 and prune while things finish before the kill
-          (shutdown_timeout/10).times do 
-            sleep (shutdown_timeout / 10.0)
-            prune_busy_threads_without_mutex
-            break if busy_size == 0
-          end
-
-          sleep(shutdown_timeout/10)
-          @busy_threads.each { |busy_thread| busy_thread.kill }
-        end
-
-        prune_busy_threads_without_mutex
+        sleep 0.01 while busy_size > 0 && Time.now.to_f < kill_at
+        @busy_threads.list.each { |busy_thread| busy_thread.kill }
       end
     end
 
     def timeout!
       return unless timeout?
 
       @mutex.synchronize do
-        @busy_threads.each do |busy_thread|
+        @busy_threads.list.each do |busy_thread|
           if (Time.now.to_i - busy_thread[:__start_time_in_seconds__] > @timeout)
             busy_thread.kill
```

Comment:
Can threads get stuck here. I wonder if it would be better to raise an error in the thread before killing it. I'm not sure if it would create more noise than help, though.